chainer/iterators/multithread_iterator.py

```diff
@@ -0,0 +1,214 @@
+from __future__ import division
+from multiprocessing import pool
+import threading
+
+import numpy
+import six
+
+from chainer.dataset import iterator
+
+
+class MultithreadIterator(iterator.Iterator):
+
+    """Dataset iterator that loads examples in parallel.
+
+    This is an implementation of :class:`~chainer.dataset.Iterator` that loads
+    examples with worker threads. It uses the standard :mod:`threading`
+    module to parallelize the loading.
+
+    Note that this iterator effectively prefetches the examples for the next
+    batch asynchronously after the current batch is returned.
+
+    This iterator saves ``-1`` instead of ``None`` in snapshots since some
+    serializers do not support ``None``.
+
+    Args:
+        dataset (~chainer.dataset.Dataset): Dataset to iterate.
+        batch_size (int): Number of examples within each batch.
+        repeat (bool): If ``True``, it infinitely loops over the dataset.
+            Otherwise, it stops iteration at the end of the first epoch.
+        shuffle (bool): If ``True``, the order of examples is shuffled at the
+            beginning of each epoch. Otherwise, examples are extracted in the
+            order of indexes.
+        n_threads (int): Number of worker threads.
+
+    """
+
+    def __init__(self, dataset, batch_size, repeat=True, shuffle=True,
+                 n_threads=1):
+        self.dataset = dataset
+        self.batch_size = batch_size
+        self._repeat = repeat
+        self._shuffle = shuffle
+        self._prefetch_order = None  # used at the end of each epoch
+
+        self.n_threads = n_threads
+
+        self._finalized = None
+
+        self.reset()
+
+    def reset(self):
+        if getattr(self, 'current_position', 0) != 0:
```

Comment:
If these attributes are initialized in the constructor, `getattr` would not be needed.